aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,180 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(DynamoLockManager.class);
+
+  private static final String LOCK_TABLE_COL_TABLE_ID = "tableId";
+  private static final String LOCK_TABLE_COL_EXPIRE_TS_MILLIS = "expireTimestampMillis";
+
+  private static final List<KeySchemaElement> LOCK_TABLE_SCHEMA = Lists.newArrayList(
+      KeySchemaElement.builder()
+          .attributeName(LOCK_TABLE_COL_TABLE_ID)
+          .keyType(KeyType.HASH)
+          .build()
+  );
+
+  private static final List<AttributeDefinition> LOCK_TABLE_COL_DEFINITIONS = Lists.newArrayList(
+      AttributeDefinition.builder()
+          .attributeName(LOCK_TABLE_COL_TABLE_ID)
+          .attributeType(ScalarAttributeType.S)
+          .build()
+  );
+
+  private final DynamoDbClient dynamo;
+  private final AwsProperties awsProperties;
+
+  DynamoLockManager(AwsProperties awsProperties) {
+    this(AwsClientUtil.defaultDynamoClient(), awsProperties);
+  }
+
+  DynamoLockManager(DynamoDbClient dynamo, AwsProperties awsProperties) {
+    this.dynamo = dynamo;
+    this.awsProperties = awsProperties;
+    ensureLockTableExists();
+  }
+
+  private void ensureLockTableExists() {
+    ensureTableExists(awsProperties.glueCatalogLockTable(), LOCK_TABLE_SCHEMA, LOCK_TABLE_COL_DEFINITIONS);
+  }
+
+  private void ensureTableExists(String tableName, List<KeySchemaElement> schema,
+                                 List<AttributeDefinition> definitions) {
+    try {
+      dynamo.describeTable(DescribeTableRequest.builder()
+          .tableName(tableName)
+          .build());
+    } catch (ResourceNotFoundException e) {
+      LOG.info("Glue lock DynamoDB table <{}> not found, trying to create", tableName);
+      dynamo.createTable(CreateTableRequest.builder()
+          .tableName(tableName)
+          .keySchema(schema)
+          .attributeDefinitions(definitions)
+          .billingMode(BillingMode.PAY_PER_REQUEST)
+          .build());
+
+      boolean isTableActive = false;
+      while (!isTableActive) {
+        LOG.info("waiting for DynamoDB table <{}> to be active", tableName);
+        try {
+          Thread.sleep(5000);
+        } catch (InterruptedException ie) {
+          LOG.warn("Glue lock DynamoDB table creation sleep interrupted", e);
+        }
+        DescribeTableResponse describeTableResponse = dynamo.describeTable(DescribeTableRequest.builder()
+            .tableName(tableName)
+            .build());
+        isTableActive = describeTableResponse.table().tableStatus().equals(TableStatus.ACTIVE);
+      }
+    }
+  }
+
+
+  @Override
+  public boolean tryLock(String database, String table, long expireMillis) {
+    String tableId = tableId(database, table);
+
+    Map<String, AttributeValue> key = Maps.newHashMap();
+    key.put(LOCK_TABLE_COL_TABLE_ID, AttributeValue.builder().s(tableId).build());
+    key.put(LOCK_TABLE_COL_EXPIRE_TS_MILLIS, AttributeValue.builder().n(
+        Long.toString(System.currentTimeMillis() + expireMillis)
```

Comment:
I think `System.currentTimeMillis` may lead to [clock skew problem](https://stackoverflow.com/questions/2978598/will-system-currenttimemillis-always-return-a-value-previous-calls/2979239#2979239). DDBLockClient's implementation to workaround this problem is interesting, probably worth borrowing: https://github.com/awslabs/amazon-dynamodb-lock-client#how-we-handle-clock-skew

Reply:
yeah I thought about that but not sure if we need to handle it given the expiration is 20 minutes or more. And I don't fully understand what that method does, it says

```
The way locks are expired is that a call to acquireLock reads in the current lock, checks the RecordVersionNumber of the lock (which is a GUID) and starts a timer. If the lock still has the same GUID after the lease duration time has passed, the client will determine that the lock is stale and expire it.
```

Does that mean the next process has to wait for the duration of expiration to acquire the lock?

Reply:
I think there's no guarantee on how far away the clock can skew, so 20min would reduce the chance but wouldn't fully avoid the problem. Regarding its explanation, I think in addition to expiration time it also has the concept of heartbeat which has a shorter duration than the expiration time, and the lock holding thread will need to heartbeat and update the GUID once a while to extend the lease, so that when another thread observes a GUID, wait for the expiration time, and come back and still see the same GUID, then that definitely means the old thread died. In this case we probably don't want to set expiration time to be as long as 20min since that would mean if a thread died even without concurrent thread waiting, the next coming thread will have to wait for 20min before considering the lock as expired.

Reply:
I see. I think this is probably a feature we should add after the initial commit of the lock interface if necessary, as long as we keep the interface consistent. Otherwise the PR will be too big. Let me add a comment for this for now.

Reply:
I think that this should be handled soon. Correctness is the most important concern and I think clock skew is a significant problem. I like the approach of storing how long other clients should wait in the lock state to handle skew. We can implement heartbeats to reduce that wait time later, but I think it is important to fix it now. Otherwise, we will need to change the lock table and that's going to be more difficult to do without downtime and across library versions.